src/core/metrics.cc

```diff
@@ -226,13 +226,17 @@ bool metric_id::operator==(
     return as_tuple() == id2.as_tuple();
 }
 
-// Unfortunately, metrics_impl can not be shared because it
-// need to be available before the first users (reactor) will call it
+shared_ptr<impl> get_local_impl(int handle) {
+    auto& impls = get_metric_implementations();
 
-shared_ptr<impl>  get_local_impl() {
-    static thread_local auto the_impl = ::seastar::make_shared<impl>();
-    return the_impl;
+    if (!impls.contains(handle)) {
+        auto ptr = ::seastar::make_shared<impl>();
+        impls[handle] = ptr;
+    }
+
+    return impls[handle];
```

Comment:
Reduce lookups from 3 to 1:
```suggestion
    auto [it, inserted] = impls.try_emplace(handle);
    if (inserted) {
        it->second = ::seastar::make_shared<impl>();
    }
    return it;
```